File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.py
class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        adj = collections.defaultdict(list)
        for u,v,c in flights:
            adj[u].append((v, c))
        vis, q = [float('inf')]*n, deque([(0, 0, src)])
        
        while q:
            k0, curr, node = q.popleft()
            if node==dst:
                vis[dst] = min(vis[dst], curr)
                continue
            if k0>k or vis[node]<curr:
                #cause all the cost less than curr has smaller k value also so that's better than this curr
                continue
            vis[node]=min(vis[node], curr)
            for v, c in adj[node]:
                q.append((k0+1, curr+c, v))
                
        return vis[dst] if vis[dst]!=float('inf') else -1
        
        
# class Solution:
#     def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
#         adj = collections.defaultdict(list)
#         for u,v,c in flights:
#             adj[u].append((v, c))
        
#         vis = [[math.inf]*(k+2) for _ in range(n)]
#         minhp = [(0, 0, src)]
        
#         while minhp:
#             curr, k0, node = heappop(minhp)
            
#             if node==dst and k0>k+1:
#                 continue
#             if node!=dst and k0>k:
#                 continue

#             i=0
#             while i<=k0:
#                 if vis[node][i] <= curr:
#                     break
#                 i+=1
#             if i!=k0+1:
#                 continue
#             vis[node][k0] = curr
            
            
#             for v, c in adj[node]:
#                 heappush(minhp, (curr+c, k0+1, v))
                
#         ans = min(vis[dst])
#         return ans if ans!=math.inf else -1
        

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        HashMap<Integer, List<int[]>> map = new HashMap<>();
        for(int[] f : flights){
            List<int[]> l = map.getOrDefault(f[0], new ArrayList<>());
            l.add(new int[]{f[1], f[2]});
            map.put(f[0], l);
        }
        Queue<int[]> q = new LinkedList<>();
        int[] vis = new int[n];
        Arrays.fill(vis, Integer.MAX_VALUE);
        q.offer(new int[]{0, 0, src});     // stops, curr distance, node
        
        while(!q.isEmpty()){
            int[] pop = q.poll();
            if(pop[2]==dst){
                vis[dst] = Math.min(vis[dst], pop[1]);
                continue;
            }
            if(pop[0]>k || vis[pop[2]]<pop[1]) continue;
            vis[pop[2]] = Math.min(vis[pop[2]], pop[1]);
            for(int i=0; i<map.getOrDefault(pop[2], new ArrayList<>()).size(); ++i){
                q.add(new int[]{pop[0]+1, pop[1]+map.get(pop[2]).get(i)[1], map.get(pop[2]).get(i)[0]});
                
            }
        }
        return vis[dst]!=Integer.MAX_VALUE ? vis[dst] : -1;
    }
}